runtime/resolvers/resource_status.go

```diff
@@ -0,0 +1,140 @@
+package resolvers
+
+import (
+	"context"
+	"errors"
+	"io"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/mitchellh/mapstructure"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/pkg/typepb"
+)
+
+func init() {
+	runtime.RegisterResolverInitializer("resource_status", newResourceStatus)
+}
+
+// resourceStatusResolver is a resolver that returns an overview of the instance's resources and their reconcile status.
+// The output fields are:
+//   - type: the resource type
+//   - name: the resource name
+//   - status: the reconcile status of the resource (one of "pending", "running", "idle")
+//   - error: the error message if the resource is in error state
+type resourceStatusResolver struct {
+	runtime    *runtime.Runtime
+	instanceID string
+	props      *resourceStatusProps
+}
+
+// resourceStatusProps declares the properties for the "resource_status" resolver.
+type resourceStatusProps struct {
+	// WhereError is a flag to only return resources that are in the error state.
+	WhereError bool `mapstructure:"where_error"`
+}
+
+// newResourceStatus creates a new resourceStatusResolver.
+func newResourceStatus(ctx context.Context, opts *runtime.ResolverOptions) (runtime.Resolver, error) {
+	props := &resourceStatusProps{}
+	if err := mapstructure.Decode(opts.Properties, props); err != nil {
+		return nil, err
+	}
+
+	return &resourceStatusResolver{
+		runtime:    opts.Runtime,
+		instanceID: opts.InstanceID,
+		props:      props,
+	}, nil
+}
+
+func (r *resourceStatusResolver) Close() error {
+	return nil
+}
+
+func (r *resourceStatusResolver) Cacheable() bool {
+	return false
+}
+
+func (r *resourceStatusResolver) Key() string {
+	hash, err := hashstructure.Hash(r.props, hashstructure.FormatV2, nil)
+	if err != nil {
+		panic(err)
+	}
+	return strconv.FormatUint(hash, 16)
+}
+
+func (r *resourceStatusResolver) Refs() []*runtimev1.ResourceName {
+	return nil
+}
+
+func (r *resourceStatusResolver) Validate(ctx context.Context) error {
+	return nil
+}
+
+func (r *resourceStatusResolver) ResolveInteractive(ctx context.Context) (runtime.ResolverResult, error) {
+	ctrl, err := r.runtime.Controller(ctx, r.instanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	rs, err := ctrl.List(ctx, "", "", false)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.props.WhereError {
+		// In-place trim resources that are not in error state
+		i := 0
+		for i < len(rs) {
+			r := rs[i]
+			if r.Meta.ReconcileError == "" {
+				// Remove from the slice
+				rs[i] = rs[len(rs)-1]
```

Comment:
We can consider `slices.Delete` ?

Reply:
Unfortunately `slices.Delete` maintains order and is therefore much slower than this, especially for repeated invocations. (Larger projects may end up with 1000s of alerts and reports, so performance is if not hugely important, at least somewhat a factor here.)